aiohttp/web_app.py

```diff
@@ -416,33 +417,102 @@ def exceptions(self) -> list[BaseException]:
 class CleanupContext(_CleanupContextBase):
     def __init__(self) -> None:
         super().__init__()
-        self._exits: list[AsyncIterator[None]] = []
+        # _exits stores either async iterators (legacy async generators)
+        # or async context manager instances. On cleanup we dispatch to
+        # the appropriate finalizer (``__anext__`` for iterators and
+        # ``__aexit__`` for context managers).
+        self._exits: list[object] = []
 
     async def _on_startup(self, app: Application) -> None:
+        """Run registered cleanup context callbacks at startup.
+
+        Each callback may return either an async iterator (an async
+        generator that yields exactly once) or an async context manager
+        (from :func:`contextlib.asynccontextmanager`). If a context manager
+        is returned it will be entered on startup (``__aenter__``) and
+        exited during cleanup (``__aexit__``). Legacy single-yield async
+        generator cleanup contexts continue to be supported.
+
+        """
         for cb in self:
-            it = cb(app).__aiter__()
-            await it.__anext__()
-            self._exits.append(it)
+            # Call the registered callback and inspect its return value.
+            # If the callback returned a context manager instance, use it
+            # directly. Otherwise (legacy async generator callbacks) we
+            # convert the callback into an async context manager and
+            # call it to obtain a context manager instance.
+            ctx = cb(app)
+
+            # If the callback returned an async iterator (legacy async
+            # generator), use it directly. If it returned an async
+            # context manager instance, enter it and remember the manager
+            # for later exit. As a final fallback, convert the callback
+            # into an async context manager (covers some edge cases) and
+            # enter that.
+            if isinstance(ctx, AsyncIterator):
+                # Legacy async generator cleanup context: advance it once
+                # (equivalent to entering) and remember the iterator for
+                # finalization.
+                it = cast(AsyncIterator[None], ctx)
+                await it.__anext__()
+                self._exits.append(it)
+            elif isinstance(ctx, contextlib.AbstractAsyncContextManager):
+                # If ctx is an async context manager: enter it and
+                # remember the manager for later exit.
+                cm = ctx
+                await cm.__aenter__()
+                self._exits.append(cm)
+            else:
+                # cb may have a broader annotated return type; adapt the
+                # callable into an async context manager and enter it.
+                cm = contextlib.asynccontextmanager(
+                    cast(Callable[[Application], AsyncIterator[None]], cb)
+                )(app)
+                await cm.__aenter__()
+                self._exits.append(cm)
```

Comment:
As I mentioned originally, can't this be reduced to just this?
```suggestion
            if not isinstance(ctx, contextlib.AbstractAsyncContextManager):
                ctx = contextlib.asynccontentmanager(cb)(app)

            await ctx.__aenter__()
            self._exits.append(ctx)
```

Reply:
I tried to reduce it to your suggested version, but I think it changes semantics in a way that breaks cleanup-context lifecycle invariants. We end up with two generator instances because we also call `ctx = cb(app)` before wrapping, the net effect is that `cb(app)` is invoked twice in the fallback path:
1- `ctx = cb(app`)  first call
2- `contextlib.asynccontextmanager(cb)(app)`  second call

For async-generator style cleanup callbacks this yields two generator instances. The instance you advance on startup is not the same instance you finalize on cleanup. 

I might have missunderstood your comment. Did you intend to change semantics so that `cb` may be invoked more than once? or did you mean to wrap the returned value when it’s not a context manager?

Reply:
> the net effect is that `cb(app)` is invoked twice

Yes, I  had realised that, but not sure it causes any actual problems. We can then probably deprecate the generator fallback in future to remove this, after a couple of years.

> The instance you advance on startup is not the same instance you finalize on cleanup.

That shouldn't be true. The second instance replaces the first one before it is used. So, if the first instance is a generator object, it gets discarded without being used at all. The second instance is then used for both startup and cleanup.

Reply:
applied your suggestion.
ready for the next review.